resolves issue #2174

Pratham1812 opened the pull request:
<!--🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅

You can expedite processing of your PR by using this template to provide context
and additional information. Before actually opening a PR please make sure that it
does NOT fall into any of the following categories

🚫 Spam PRs (accidental or intentional) - these will result in a 7 / 30 / ∞ days ban from
interacting with the project depending on reoccurrence and severity. You can find more
information [here](https://pwning.owasp-juice.shop/companion-guide/latest/part3/contribution.html#_handling_of_spam_prs). 

🚫 Lazy typo fixing PRs - if you fix a typo in a file, your PR will only be merged
if all other typos in the same file are also fixed with the same PR

🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅🔅-->

### Description

This PR adds a feature to manually disable a "danger zone" challenge even if the running environment is not containerised or the container environment is any other alternative to "Docker"


Resolved or fixed issue: #2174

### Affirmation

- [x] My code follows the [CONTRIBUTING.md](https://github.com/juice-shop/juice-shop/blob/master/CONTRIBUTING.md) guidelines

bkimminich commented:
Hi! Thanks for the PR. There are several issues with it, though. Here's why this does not suffice as a solution for #2174:
⛔ Two opposing boolean flags instead of one consistent enum to configure the behavior
⛔ [Score Board frontend](https://github.com/juice-shop/juice-shop/tree/develop/frontend/src/app/score-board) does not reflect the new configuration options
⛔ PR sent towards master branch instead of develop (where quite a few refactorings happened in the meantime, including a removal of the legacy Score Board

bkimminich commented:
I've converted the PR to draft as it contains linting errors. You can run `npm run lint` locally before pushing commits to avoid these from just popping up during CI/CD for the first time. Also, the introduced enum currently only supports "on" and "off" but not the "auto" behavior that would be the default. See https://github.com/juice-shop/juice-shop/issues/2174#issuecomment-1883766917.

Also, the Score Board does not reflect the new behavior where dangerous challenges could be disabled independent of environment. Here's an issue with the current implementation:

1. Juice Shop is running on a Linux machine locally, not in a container
2. Consider safety mode is "on"
3. User goes to Score Board. They see all challenges as available, but the dangerous ones should be disabled

Pratham1812 commented:
Hey I did the relevant changes, also tested on docker environment for "auto" functionality. Currently I have configured it such that if running on docker container with safety mode flag switched to "off",I am still disabling challenges which were meant to be disabled on docker so as to not break the current flow, should anything else be done?

bkimminich commented:
| `safetyMode` | Behavior on Docker/Heroku/GitPod  | Behavior on Windows | PR behavior matches? |
|--------|--------|--------|--------|
| `auto` | Dangerous challenges disabled | Dangerous enabled; broken JWT Challenge disabled* | ❓|
| `off` | Dangerous challenges enabled | Dangerous enabled; broken JWT Challenge disabled* | ❓ |
| `on` | Dangerous challenges disabled | Dangerous disabled; broken JWT Challenge disabled* | ❓ |

Optimally, you should have an automated test case (in the API test suite probably) that checks this behavior.

*this could only be achieved, if the check for the broken JWT challenge on Windows would be decoupled from the safety mode mechanism entirely. If it stays in the hands of the mechanism, then `auto` would still disable the broken JWT challenge if Windows is detected and `on` would always disable it.

Pratham1812 commented:
Hey, I have fixed the functionality, could you please check , I am attaching the screenshots for your refference, which I had taken replicating different environments(windows & docker)

![win](https://github.com/juice-shop/juice-shop/assets/32198580/40e39874-e9ad-44c7-9812-b59a6027cd7d)
windows with `safetyMode:off`
![docker_auto](https://github.com/juice-shop/juice-shop/assets/32198580/a73b372f-42e3-44c2-bed0-faad453c2cda)
docker environment with `safetyMode:auto`

I ran the code by running `npm i && npm run start`

bkimminich commented:
Thanks, I'll do a full local test in the next couple days! 👍

Pratham1812 commented:
Hey, did you test the changes?

bkimminich commented:
Not yet, it's on my todo list but due to an upcoming business trip it might take a few more days, sorry.

bkimminich reviewed (commented):
## Tooltip no longer shows environment/OS
![image](https://github.com/juice-shop/juice-shop/assets/3531020/4efe474f-5e8b-4c0b-b4cb-431669daca37)

This previously would show the name of the environment/OS that caused the challenge to be disabled, e.g. the above should be "...in a **Windows** environment!"

## With `safetyMode: on` the tooltip should be more explicit
When safety mode is turned on, the dangeorus challenges should always be disabled, but the messages on the Score Board should make clear that this is a hard configuration and not depending on any environment.

## Unsafe config file now behaves as default
Currently is 
```
challenges:
  safetyMode: auto
```
but must be
```
challenges:
  safetyMode: off
```

Pratham1812 commented:
Hey, can you provide me with appropriate text which must be displayed in case of `safetyMode : on` so that I can add appropriate translation for the text in all files of `frontend/src/assets/i18n/`

bkimminich commented:
> Hey, can you provide me with appropriate text which must be displayed in case of `safetyMode : on` so that I can add appropriate translation for the text in all files of `frontend/src/assets/i18n/`

How about this: "This challenge is unavailable because the application is running in safety mode!"

But for `auto`-mode it must remain the text as it was with the environment being mentioned.

Pratham1812 commented:
Hey @bkimminich, what other issues can I take up? Also will juice shop be part of gaon this year?

bkimminich commented:
> Hey @bkimminich, what other issues can I take up? Also will juice shop be part of gsoc this year?

Hi! I went over your solution again and again, but I'm honestly not sure if it can be considered mergable at this point. The behavior might be as intended, but the solution is incredible brittle and makes an originally sub-optimal approach even more confusing 🍝 ... Just one example:

<img width="983" alt="image" src="https://github.com/juice-shop/juice-shop/assets/3531020/dda30cb2-5afc-471e-9ad5-2d5c3a027173">

I'm leaning towards re-implementing the entire mechanism from scratch, but that requires some time I currently do not have. I'm not 100% sure yet if the current codebase or your PR is a better starting point for a re-implementation, but I'm leaning towards not merging the PR.

@J12934 @ShubhamPalriwala, what do you guys think?

Pratham1812 commented:
Hey, I realise the code quality is not good for merging, but considering that I know the pain points and how to implement the necessary functionality, Can I try for reimplementation of the code? and show you the results? Also if any other suggestion that you would like to add.

J12934 reviewed (commented):
My suggestions for a refactoring here would be:

- replace `disableOnContainerEnv` & `disableOnWindowsEnv` & `disableOnFlagSet` with a `function isChallengeEnabled (challenge: Challenge): boolean`. E.g. used as `if (utils.isChallengeEnabled(challenges.httpHeaderXssChallenge)) { ... }`. This should then look at the challenge and the current environment and juice shop config and return true or false if the challenge should be disabled. This means we would only have to do one check / function call in the individual route. Would be good if this `isChallengeEnabled` would have tests to check if challenges are properly enabled/disabled on the respective environments (using mocked versions of the is-docker, is-windows... functions and `safetyMode` configurations).
- for the cypress tests this might be a bit hard with the tasks, i don't know how easily we could pass them a challenge object. I'd suggest using the is-docker, is-windows functions there instead for the individual checks.
- the config value should be validated in `lib/startup/validateConfig.ts` and should stop the startup with an helpful error message when the safetyMode is set to un unexpected value. (not one of the three expected strings)
- (not a refactoring but a change in requirements): I'd drop the added text to the description of a disabled / "enabled but dangerous" challenge. It's already quite clear in the score board for both of them I'd say.

Pratham1812 commented:
Hey, I have improved the code with the existing functionality.As pointed by @J12934 , I have updated `validateconfig.ts` and instead of using `on`,`off` and `auto`, I am using `enabled`,`disabled`,`auto`. I have changed the way the disabledEnv is getting set for each challenge and also simplified the approach of disabling necessary challenges.
